### PR TITLE
Enable dispatch to tinygemm int4 and int8 kernels for quantized tensor

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
This adds some dispatch to the tinygemm kernels for cuda, although need to resolve implementation mismatch problem for tinygemm first

Test Plan:
TODO

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by jerryzh168** [comment]
waiting for https://github.com/pytorch/ao/pull/227 to be landed to actually test it

File: torchao/quantization/subclass.py [hunk 1]
```diff
@@ -832,7 +836,54 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
                     args[1],
                     None if len(args) == 2 else args[2],
                 )
-            if weight_qtensor.input_quant_func is not None:
+            if weight_qtensor.input_quant_func is None:
```
Comment (cpuhrsch) [inline torchao/quantization/subclass.py:839]: Let's please remove `input_quant_func` and discuss using `AffineQuantizedTensor` as an organizing principle :)

Comment (jerryzh168) [reply torchao/quantization/subclass.py:839]: this is used for testing 8da4w right now, can this be done later until there is a better alternative?

**Review by cpuhrsch — APPROVED** [review]

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Setting request changes just so we don't forget to remove `input_quant_func`

**Comment by jerryzh168** [comment]
> Setting request changes just so we don't forget to remove `input_quant_func`

will do in next PR

File: torchao/quantization/subclass.py [hunk 2]
```diff
@@ -865,3 +917,43 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
                 kwargs,
                 args[0].to(*args[1:], **kwargs)._apply_fn_to_data(torch.clone),
             )
+
+
+# TODO: add padding support
+class TinygemmAffineQuantizedTensor(AffineQuantizedTensor):
```
Comment (cpuhrsch) [inline torchao/quantization/subclass.py:923]: Why this?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
CI is red and let's please not add a TinygemmAffineQuantizedTensor

**Comment by jerryzh168** [comment]
sorry, CI should be fixed now.

For tinygemm, it's because it is quantizing things differently, we need to think of how to unify this

**Review by cpuhrsch — APPROVED** [review]

File: test/quantization/test_quant_api.py [hunk 3]
```diff
@@ -50,14 +49,14 @@ def _apply_dynamic_quant(model):
     """
     _replace_with_custom_fn_if_matches_filter(
         model,
-        lambda linear_mod: dynamic_quant(linear_mod, (torch.randn(1, linear_mod.in_features))),
+        lambda linear_mod: dynamic_quant(linear_mod, (torch.randn(1, linear_mod.in_features),)),
```
Comment (cpuhrsch) [inline test/quantization/test_quant_api.py:52]: Interesting. Why is that extra comma needed now?

File: test/quantization/test_quant_primitives.py [hunk 4]
```diff
@@ -351,16 +353,11 @@ def test_tinygemm_get_groupwise_affine_qparams(self):
...
 
-        def int_zero_point_to_float(zero_point, scale, qaunt_min, mid_point):
-            return (quant_min - zero_point + mid_point) * scale
-
-        mid_point = 2 ** (n_bit - 1)
-        zero_point_float = int_zero_point_to_float(zero_point, scale, quant_min, mid_point)
-
         self.assertTrue(torch.equal(scale, scale_ref))
-        torch.testing.assert_close(zero_point_float, zero_point_ref, rtol=0.00001, atol=torch.max(scale)*0.03)
+        self.assertTrue(torch.equal(zero_point, zero_point_ref))
```
Comment (cpuhrsch) [inline test/quantization/test_quant_primitives.py:360]: Neat

File: test/quantization/test_quant_api.py (hunk 3 above)
Comment (cpuhrsch) [reply test/quantization/test_quant_api.py:52]: By all means don't be blocked on this comment haha

**Pull request merged**